worker/caasadmission/k8s/admission.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package k8s
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"github.com/juju/errors"
+	admission "k8s.io/api/admissionregistration/v1beta1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/juju/juju/caas/kubernetes/provider"
+)
+
+type CertificateBroker interface {
+	// Register a new domain name lease on the brokers certificate. Returns a
+	// lease function if no errors
+	RegisterDomainLeases(...string) (func(), error)
+
+	// Set of root certificates in pem format
+	RootCertificates() ([]byte, error)
+}
+
+// TODO this is a tmp struct and won't make it to final
+type CertWatcherBroker struct {
+	CertWatcher func() *tls.Certificate
+}
+
+type K8sBroker interface {
+	EnsureMutatingWebhookConfiguration(*admission.MutatingWebhookConfiguration) (func(), error)
+}
+
+// Represents a creator of mutating webhooks that is context aware of the
+// current controller.
+type AdmissionCreator interface {
+	EnsureMutatingWebhookConfiguration() (func(), error)
+}
+
+// Func type of AdmissionCreator
+type AdmissionCreatorFunc func() (func(), error)
+
+const (
+	Component = "admission"
+)
+
+var (
+	anyMatch = []string{"*"}
+)
+
+// Implements AdmissionCreator interface for func type
+func (a AdmissionCreatorFunc) EnsureMutatingWebhookConfiguration() (func(), error) {
+	return a()
+}
+
+// Instantiates a new AdmissionCreator for the supplied context arguments.
+func NewAdmissionCreator(
+	certBroker CertificateBroker,
+	namespace, modelName, modelUUID string,
+	ensureConfig func(*admission.MutatingWebhookConfiguration) (func(), error),
+	service *admission.ServiceReference) (AdmissionCreator, error) {
+
+	caPems, err := certBroker.RootCertificates()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	failurePolicy := admission.Ignore
```

Comment:
Did we discuss using admission.Fail?
What's the plan for dealing with resources without juju labels if said resources are created when the controller is not running?

Reply:
Failure policy handles two cases. It defines what happens if the server 500's or the server is not available. We did discuss making it fail but I left it as ignore for now so I didn't break anything due to buggy code. Plan is to swap it.

I'll add a TODO.